tree/tree/src/TBranchElement.cxx

```diff
@@ -2244,6 +2250,59 @@ TClass* TBranchElement::GetCurrentClass()
    return cl;
 }
 
+void TBranchElement::InitializeBranchLists()
+{
+   const auto nbranches = fBranches.GetEntriesFast();
+
+   // If the branch holds a class featuring schema rules, we cannot parallelise
+   // since a certain order is to be imposed
+   if(auto heldClass = GetClass()) { // perhaps always true
+      if (heldClass->GetSchemaRules()) {
+         // we just put all the branches in the set to be read sequentially
+         for (Int_t i = 0; i < nbranches; i++)  {
+            fSeqBranches.emplace_back((TBranch*)fBranches.UncheckedAt(i));
+         }
+         return; // nothing else to be done
+      }
+   }
+
+   // The branches to be processed sequentially are those that are the leaf count of another branch
```

Comment:
By Processing the 'rules' affected branches and the leafcount branch separately, technically the ordering is changed (possibly fatally) .... except that for now the test (line 2260) is applied equally to all members of the class so the cases of problem will be rare (eg. split object, leafcount in base class, derived class with both using the base class leafcount and have a read-rule).

Reply:
I am not sure I get this. If there is at least a rule, all branches are processed sequentially. The leaf counts are treated as in the TTree::GetEntry case. To be even sturdier, at the cost of some parallelisation potential, I could detect leafcounts, if any, process the subbranches sequentially.

Reply:
The order of branch can be:
1. base class with no rules, leafcount
1. derived class with rules, member1
1. same derived class with rules, member2, variable size array using the base class leafcount.

if I understood correctly the algorithm, 2 and 3 will be processed sequentially and 1 will also be processed sequentially *but* in the following order:  2 then 3 then 1 ....

Reply:
It could be that I do not understand the comment, but I think that, in the current implementation, if the class has rules, the behaviour is the current, sequential one.

Reply:
Yes but in the leaves are processed in the wrong order in the example I gave.
```
class Base {
   Int_t fN;
};
class Derived : public Base {
   Float_t *fArray; //[fN]
   Double_t fWithRules; // This member and only this member has a rules.
};
```
lead to the branches
```
top.fN  -> associated with class Base
top.fArray -> associated with class Derived
top.fWithRules -> Associtated with class Derived.
```
and if I read the code correctly this will lead to the 'fSeqBranches' collection to hold a pointer to the branch in the following *incorrect* order
```
top.fArray -> put there by 1st loop
top.fWithRules -> out there by 1st loop.
top.fN -> put there by the 2nd loop find leafcount.
```
The two loops putting stuff in the sequential list must be merged.

Reply:
I see. But I think we check for rules. If they are there, we put all branches in the set of branches to be processed sequentially. We return (https://github.com/dpiparo/root/blob/ffce3ec45b222095ff9dfed76de18d2f370e3867/tree/tree/src/TBranchElement.cxx#L2265). At that point, isn't the behaviour identical to the sequential one?

Reply:
> At that point, isn't the behaviour identical to the sequential one?

I see ... I had not noticed the return statement.  If that return statement is taken, fSortedBranches is empty but it seems there is then a 'spurrious' execution of pool.Foreach.